src/server/zset_family.cc

```diff
@@ -2355,12 +2440,70 @@ void ZSetFamily::ZInterCard(CmdArgList args, const CommandContext& cmd_cntx) {
   builder->SendLong(result.value().size());
 }
 
+void ZSetFamily::ZMPop(CmdArgList args, const CommandContext& cmd_cntx) {
+  uint32 num_keys;
+  bool is_max;
+  int pop_count;
+  if (!ValidateZMPopCommand(args, &num_keys, &is_max, &pop_count, cmd_cntx.rb)) {
+    return;
+  }
+  auto* response_builder = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
+
+  // From the list of input keys, keeps the first (in the order of keys in the command) key found in
+  // the current shard.
+  absl::flat_hash_set<std::string> first_found_key_in_shard;
+  // We can have at most one result from each shard.
+  first_found_key_in_shard.reserve(std::min(shard_set->size(), num_keys));
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    std::optional<std::string> result = GetFirstNonEmptyKeyFound(shard, t);
+    if (result.has_value()) {
+      first_found_key_in_shard.insert(std::move(*result));
+    }
+    return OpStatus::OK;
+  };
+
+  cmd_cntx.tx->Execute(std::move(cb), false /* possibly another hop */);
+
+  // Now that we have the first non empty key from each shard, find the first overall first key and
+  // pop elements from it.
+
+  auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
+
+  std::string_view key_to_pop;
+  ArgRange arg_keys(args.subspan(1, num_keys));
+  // Find the first arg_key which exists in any shard and is not empty.
+  for (std::string_view key : arg_keys) {
+    if (first_found_key_in_shard.contains(key)) {
+      key_to_pop = key;
+      break;
+    }
+  }
+
+  if (key_to_pop.empty()) {
```

Comment:
Empty string is also a valid key
you can run 
zadd "" 1 f1

Reply:
Done